src/rebar_compiler.erl

```diff
@@ -115,7 +115,7 @@ clean(Compilers, AppInfo) ->
 %% These functions are here for the ultimate goal of getting rid of
 %% rebar_base_compiler. This can't be done because of existing plugins.
 
--spec needs_compile(filename:all(), extension(), [{extension(), file:dirname()}]) -> boolean().
+-spec needs_compile(file:name_all(), extension(), [{extension(), string()}]) -> boolean().
```

Comment:
those should all be file paths, and I'm not the biggest fan of getting just "string()" there. What kind of error were you getting?

Reply:
Types are not exported. I can name them, if that makes a difference.
Otherwise we can get rid of `unknown`, in `rebar.config` and get over a whole class of static analysis issues.

Reply:
I mean not just these, the file overall has a lot of filename renamed to strings; we should just use the proper filename types when possible.

Reply:
You mean "use them albeit not exported"? (I don't mind reverting the changes, but I tend to use only exported types)
How about creating a local type for those (for naming sake) and making that a `string()`?

Reply:
I see several options here:
1. use non-exported types (as-was) - and let `dialyzer` complain away - as long as they're declared internally (in Erlang/OTP), and ignore the warnings (I propose we get rid of `unknown`, in this case, since supposedly tests should surface any issues with `rebar`),
2. create internal types, for semantics purposes, to not have `string` everywhere we shouldn't (e.g. `file_name_all()`) - not all are `string()` since I evaluated every situation individually,
3. name the variables, like `Filename :: string()`, so stuff doesn't lose meaning.

There're always tradeoffs, I guess, but ultimately it's up to you to decide. I personally prefer to not have warnings (as much as possible - since amid some of those there's actually valuable stuff that might go unnoticed).

Reply:
Given the usage: https://erlang.org/doc/man/file.html#data-types I'd say `file:filename()`

Reply:
I've replaced the occurrences I'd replaced in `rebar_compiler.erl` and `rebar_file_utils.erl` (now force-pushed).